### ISSUE TITLE
Build error in pmix_client_get

### ISSUE THREAD

**Issue by lyu** [openpmix/openpmix#2684]
### What version of the PMIx Reference Library are you using?
v4.2 branch HEAD

### Describe how PMIx was installed
From a git clone

### Please describe the system on which you are running

* Operating system/version: Rocky Linux 8.4 aarch64 (GCC 12.1)

-----------------------------

## Details of the problem

My system runs GCC 12.1. When building PMIX 4.2, I encounter the following errors: (master branch builds fine)

```
client/pmix_client_get.c:84:53: error: argument 2 of type 'const char[512]' with mismatched bound [-Werror=array-parameter=]
   84 |                                    const pmix_key_t key,
      |                                    ~~~~~~~~~~~~~~~~~^~~
In file included from client/pmix_client_get.c:23:
../include/pmix.h:206:72: note: previously declared as 'const char[]'
  206 | PMIX_EXPORT pmix_status_t PMIx_Get(const pmix_proc_t *proc, const char key[],
      |                                                             ~~~~~~~~~~~^~~~~
client/pmix_client_get.c: In function 'PMIx_Get_nb':
client/pmix_client_get.c:364:26: error: '_getfn_fastpath' reading 512 bytes from a region of size 11 [-Werror=stringop-overread]
  364 |                     rc = _getfn_fastpath(&p, PMIX_HOSTNAME, &optional, 1, &ival);
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
client/pmix_client_get.c:364:26: note: referencing argument 2 of type 'const char[512]'
client/pmix_client_get.c:364:26: note: referencing argument 3 of type 'const pmix_info_t[0]' {aka 'const struct pmix_info[]'}
client/pmix_client_get.c:927:22: note: in a call to function '_getfn_fastpath'
  927 | static pmix_status_t _getfn_fastpath(const pmix_proc_t *proc, const pmix_key_t key,
      |                      ^~~~~~~~~~~~~~~
client/pmix_client_get.c:491:26: error: '_getfn_fastpath' reading 512 bytes from a region of size 12 [-Werror=stringop-overread]
  491 |                     rc = _getfn_fastpath(&pmix_globals.myid, PMIX_APPNUM, &optional, 1, &ival);
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
client/pmix_client_get.c:491:26: note: referencing argument 2 of type 'const char[512]'
client/pmix_client_get.c:491:26: note: referencing argument 3 of type 'const pmix_info_t[0]' {aka 'const struct pmix_info[]'}
client/pmix_client_get.c:927:22: note: in a call to function '_getfn_fastpath'
  927 | static pmix_status_t _getfn_fastpath(const pmix_proc_t *proc, const pmix_key_t key,
      |                      ^~~~~~~~~~~~~~~
client/pmix_client_get.c:523:22: error: '_getfn_fastpath' reading 512 bytes from a region of size 12 [-Werror=stringop-overread]
  523 |                 rc = _getfn_fastpath(&pmix_globals.myid, PMIX_APPNUM, &optional, 1, &ival);
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
client/pmix_client_get.c:523:22: note: referencing argument 2 of type 'const char[512]'
client/pmix_client_get.c:523:22: note: referencing argument 3 of type 'const pmix_info_t[0]' {aka 'const struct pmix_info[]'}
client/pmix_client_get.c:927:22: note: in a call to function '_getfn_fastpath'
  927 | static pmix_status_t _getfn_fastpath(const pmix_proc_t *proc, const pmix_key_t key,
      |                      ^~~~~~~~~~~~~~~
client/pmix_client_get.c:364:26: error: '_getfn_fastpath' reading 512 bytes from a region of size 11 [-Werror=stringop-overread]
  364 |                     rc = _getfn_fastpath(&p, PMIX_HOSTNAME, &optional, 1, &ival);
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
client/pmix_client_get.c:364:26: note: referencing argument 2 of type 'const char[512]'
client/pmix_client_get.c:364:26: note: referencing argument 3 of type 'const pmix_info_t[0]' {aka 'const struct pmix_info[]'}
client/pmix_client_get.c:927:22: note: in a call to function '_getfn_fastpath'
  927 | static pmix_status_t _getfn_fastpath(const pmix_proc_t *proc, const pmix_key_t key,
      |                      ^~~~~~~~~~~~~~~
client/pmix_client_get.c:491:26: error: '_getfn_fastpath' reading 512 bytes from a region of size 12 [-Werror=stringop-overread]
  491 |                     rc = _getfn_fastpath(&pmix_globals.myid, PMIX_APPNUM, &optional, 1, &ival);
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
client/pmix_client_get.c:491:26: note: referencing argument 2 of type 'const char[512]'
client/pmix_client_get.c:491:26: note: referencing argument 3 of type 'const pmix_info_t[0]' {aka 'const struct pmix_info[]'}
client/pmix_client_get.c:927:22: note: in a call to function '_getfn_fastpath'
  927 | static pmix_status_t _getfn_fastpath(const pmix_proc_t *proc, const pmix_key_t key,
      |                      ^~~~~~~~~~~~~~~
client/pmix_client_get.c:523:22: error: '_getfn_fastpath' reading 512 bytes from a region of size 12 [-Werror=stringop-overread]
  523 |                 rc = _getfn_fastpath(&pmix_globals.myid, PMIX_APPNUM, &optional, 1, &ival);
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
client/pmix_client_get.c:523:22: note: referencing argument 2 of type 'const char[512]'
client/pmix_client_get.c:523:22: note: referencing argument 3 of type 'const pmix_info_t[0]' {aka 'const struct pmix_info[]'}
client/pmix_client_get.c:927:22: note: in a call to function '_getfn_fastpath'
  927 | static pmix_status_t _getfn_fastpath(const pmix_proc_t *proc, const pmix_key_t key,
      |                      ^~~~~~~~~~~~~~~
client/pmix_client_get.c:364:26: error: '_getfn_fastpath' reading 512 bytes from a region of size 11 [-Werror=stringop-overread]
  364 |                     rc = _getfn_fastpath(&p, PMIX_HOSTNAME, &optional, 1, &ival);
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
client/pmix_client_get.c:364:26: note: referencing argument 2 of type 'const char[512]'
client/pmix_client_get.c:364:26: note: referencing argument 3 of type 'const pmix_info_t[0]' {aka 'const struct pmix_info[]'}
client/pmix_client_get.c:927:22: note: in a call to function '_getfn_fastpath'
  927 | static pmix_status_t _getfn_fastpath(const pmix_proc_t *proc, const pmix_key_t key,
      |                      ^~~~~~~~~~~~~~~
client/pmix_client_get.c:491:26: error: '_getfn_fastpath' reading 512 bytes from a region of size 12 [-Werror=stringop-overread]
  491 |                     rc = _getfn_fastpath(&pmix_globals.myid, PMIX_APPNUM, &optional, 1, &ival);
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
client/pmix_client_get.c:491:26: note: referencing argument 2 of type 'const char[512]'
client/pmix_client_get.c:491:26: note: referencing argument 3 of type 'const pmix_info_t[0]' {aka 'const struct pmix_info[]'}
client/pmix_client_get.c:927:22: note: in a call to function '_getfn_fastpath'
  927 | static pmix_status_t _getfn_fastpath(const pmix_proc_t *proc, const pmix_key_t key,
      |                      ^~~~~~~~~~~~~~~
client/pmix_client_get.c:523:22: error: '_getfn_fastpath' reading 512 bytes from a region of size 12 [-Werror=stringop-overread]
  523 |                 rc = _getfn_fastpath(&pmix_globals.myid, PMIX_APPNUM, &optional, 1, &ival);
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
client/pmix_client_get.c:523:22: note: referencing argument 2 of type 'const char[512]'
client/pmix_client_get.c:523:22: note: referencing argument 3 of type 'const pmix_info_t[0]' {aka 'const struct pmix_info[]'}
client/pmix_client_get.c:927:22: note: in a call to function '_getfn_fastpath'
  927 | static pmix_status_t _getfn_fastpath(const pmix_proc_t *proc, const pmix_key_t key,
      |                      ^~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

**Comment by rhc54** [comment]
I would suggest telling the compiler to take a powder - it is complaining about things that are effectively typedef'd to each other. At a practical level, all I can suggest is that you not treat warnings as errors as these are things we cannot/won't "fix".

**Comment by lyu** [comment]
OK, I have added the corresponding `-Wno-error=...` CFLAGS

**Comment by rhc54** [comment]
Just for the fun of it, I installed GCC 12.1 on my Mac and built PMIx head of master branch - and didn't get any of those warning/errors. I did finally hit some similar complaints in the test area (test/simple/gwtest.c), but nothing in the main code base. This was with `--enable-devel-check` set in configure, which turns on the picky compiler settings.

I then re-ran it on v4.2 and was able to reproduce what you saw. Of course, client_get is vastly different between v4.2 and master branches, which probably explains the difference.

I'll probably take a look at it to ensure these aren't something that could trash the stack, for example, so I'll reopen this for now.
